google/datalab/contrib/pipeline/_pipeline.py

```diff
@@ -142,14 +142,24 @@ def _get_operator_definition(self, task_id, task_details):
     operator_param_values = Pipeline._get_operator_param_name_and_values(
         operator_classname, task_details)
     for (operator_param_name, operator_param_value) in sorted(operator_param_values.items()):
+      operator_param_value = self._resolve_airflow_macros(operator_param_value)
       param_format_string = Pipeline._get_param_format_string(
           operator_param_value)
       param_string = param_format_string.format(operator_param_name, operator_param_value)
-      param_string = param_string % self._airflow_macros
       full_param_string = full_param_string + param_string
 
     return '{0} = {1}({2}, dag=dag)\n'.format(task_id, operator_classname, full_param_string)
 
+  def _resolve_airflow_macros(self, operator_param_value):
+    if type(operator_param_value) is list:
+      return [self._resolve_airflow_macros(item) for item in operator_param_value]
+    if type(operator_param_value) is dict:
+      return dict((self._resolve_airflow_macros(k), self._resolve_airflow_macros(v))
+                  for k, v in operator_param_value.items())
+    if isinstance(operator_param_value, basestring):
```

Comment:
All the other files in pydatalab that use basestring include this line:
from past.builtins import basestring

Reply:
Ah, thanks! Looks like while it's not a requirement for py2, it's indeed a requirement for py3.

Reply:
Not sure "isinstance(operator_param_value, basestring)" covers all string types in py3.

I would suggest "isinstance(operator_param_value, six.string_types)", which seems popular. Within pydatalab, see https://github.com/googledatalab/pydatalab/search?utf8=%E2%9C%93&q=six.string_types&type=.

Reply:
Done.